File: Algorithms/Swift/Tree/102_Binary_Tree_Level_Order_Traversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func levelOrder(root: TreeNode?) -> [[Int]] {
        var values = [[Int]]()

        if root == nil {
            return values
        }

        var queue = [TreeNode?]()
        queue.append(root);

        while !queue.isEmpty {
            var levelValues = [Int]()
            let levelSize = queue.count

            for var i = 0; i < levelSize; i++ {
                let node = queue.first!
                queue.removeFirst()

                levelValues.append(node!.val)

                if let left = node!.left {
                    queue.append(left)
                }
                if let right = node!.right {
                    queue.append(right)
                }
            }

            values.append(levelValues)
        }

        return values
    }
}

File: Algorithms/JavaScript/Easy/102_Binary_Tree_Level_Order_Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    var levelsValues = []

    if (root === null) {
        return levelsValues
    }

    var queue = [root]
    while (queue.length > 0) {
        var levelValues = []

        var queueLength = queue.length
        for (var i = 0; i < queueLength; i++) {
            var node = queue.shift()
            levelValues.push(node.val)

            if (node.left !== null) {
                queue.push(node.left)
            }
            if (node.right !== null) {
                queue.push(node.right)
            }
        }

        levelsValues.push(levelValues)
    }

    return levelsValues
};
